---
title: "37C Time Course Analysis"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE)
library(tidyverse)
library(ggplot2)
library(gtools)
library(stringr)
library(cowplot)
library(readxl)
SAVE_FIGS <- TRUE

cbPalette <- c("#CC79A7", "#56B4E9", "#E69F00", "#F0E442", "#0072B2", "#D55E00", "#009E73", "#999999")

# These are genes that are not included in the simulation and should be removed before comparing
# simulations with experimental data
IGNORE_GENES <- c("gene 10B",
                "possible gene 5.5-5.7",
                "gene 1.5",
                "gene 1.6",
                "gene 4.1",
                "gene 4B",
                "gene 0.6A",
                "gene 0.6B",
                "possible gene 0.6B",
                "gene 0.5",
                "gene 0.4")
```

## Tools

Reads were mapped with HISAT2 and counted with bedtools multicov. Counts were normalized and given as TPM values.

## Normalized counts show full coverage of genome

In the plot below, reads are counted towards all genes that they map to if a read overlaps with multiple genes.

```{r, echo=FALSE, fig.width = 10, fig.height = 18, fig.width = 12}

id_map <- read_csv("../output/id_map.csv") %>% filter(!is.na(gene_number), !(pinetree %in% IGNORE_GENES))

transcripts <- read_csv("../output/transcripts_37C.csv") %>% 
  filter(!is.na(class), !(pinetree %in% IGNORE_GENES))

gene_labels <- unique(mixedsort(id_map$gene_number))

rna_9_dist <- ggplot(transcripts %>% filter(time == 9), aes(x = factor(gene_number, levels=gene_labels), y = tpm)) + 
  geom_bar(stat="summary", fun.y = "mean", aes(fill = factor(class))) +
  geom_point(alpha = 0.5) + 
  theme(axis.text.x=element_blank(), axis.ticks.x=element_blank(), legend.position = c(0.0, 0.8), legend.title = element_blank(), legend.spacing.x = unit(2, 'pt')) +
  scale_fill_manual(values=cbPalette, breaks=c("1","2","3"), labels=c("class I", "class II", "class III")) +
  geom_vline(xintercept = (match(c("3.8", "4.7", "6.5", "18.5"), gene_labels)) - 0.5, linetype="dashed", color = "gray40") +
  geom_vline(xintercept = (match(c("11"), gene_labels)) - 0.5) +
  xlab("gene") + ylab("transcripts per million") + scale_y_continuous(expand = c(0,0)) +
  coord_cartesian(clip = "off")

rna_5_dist <- ggplot(transcripts %>% filter(time == 5), aes(x = factor(gene_number, levels=gene_labels), y = tpm)) + 
  geom_bar(stat="summary", fun.y = "mean", aes(fill = factor(class))) +
  geom_point(alpha = 0.5) + 
  geom_vline(xintercept = (match(c("3.8", "4.7", "6.5", "18.5"), gene_labels)) - 0.5, linetype="dashed", color = "gray40") +
  geom_vline(xintercept = (match(c("11"), gene_labels)) - 0.5) +
  theme(axis.text.x=element_blank(), axis.ticks.x=element_blank(), legend.position = c(0.0, 0.8), legend.title = element_blank(), legend.spacing.x = unit(2, 'pt')) +
  scale_fill_manual(values=cbPalette, breaks=c("1","2","3"), labels=c("class I", "class II", "class III")) +
  xlab("gene") + ylab("transcripts per million") + scale_y_continuous(expand = c(0,0)) +
  coord_cartesian(clip = "off")

rna_5_9_exp <- plot_grid(rna_5_dist, rna_9_dist, ncol = 1, align='v', labels="AUTO")

rna_5_9_exp

if (SAVE_FIGS) {
  save_plot("./figures/rna_5_9_exp.pdf", rna_5_9_exp, base_width = 8, base_height = 6)
}

```

### Individually-mapped read counts drop-off around RNAse cleavage sites and terminators

```{r}

coverage <- read_tsv("../data/rna_seq/37C/coverage/PA54_coverage_q60.tsv", col_names = c("genome", "position", "coverage"))

# Coverage map of full genome
full_map <- ggplot(coverage, aes(x = position, y = coverage)) + 
  geom_bar(stat="identity", fill="grey60") + 
  geom_vline(xintercept = 24162) + xlim(0,40000) +
  scale_y_continuous(expand = c(0,0))
# Coverage map of region around phi 6.5
rnase_6_5 <- ggplot(coverage, aes(x = position, y = coverage)) + 
  geom_bar(stat="identity", fill="grey60") + 
  geom_vline(xintercept = 18545) +
  geom_vline(xintercept = 18563, linetype="dashed") + 
  xlim(18400,18900) +
  scale_y_continuous(expand = c(0,0))
# Coverage of map of region around phi 3.8
rnase_3_8 <- ggplot(coverage, aes(x = position, y = coverage)) + 
  geom_bar(stat="identity", fill="grey60") +
  geom_vline(xintercept = 11180) + 
  geom_vline(xintercept = 11203, linetype="dashed") + 
  xlim(11050,11550) +
  scale_y_continuous(expand = c(0,0)) +

bottom_row <- plot_grid(rnase_3_8, rnase_6_5, labels = c("B", "C"))
cov_map <- plot_grid(full_map, bottom_row, ncol = 1, labels = c("A", ""))

if(SAVE_FIGS == TRUE) {
  save_plot("./figures/cov_map.pdf", cov_map, base_width = 8, base_height = 5)
}

```

### Simulated transcript distribution in wildtype

```{r}

labels <- read_csv("../output/id_map.csv") %>% filter(!is.na(gene_number))
load_sim <- function(path) {
  read_tsv(path) %>% group_by(time) %>% nest() %>%
    mutate(data = purrr::map(data, ~left_join(labels, ., by = c("pinetree" = "species")))) %>%
    unnest() %>%
    mutate(time = round(time)) %>%
    mutate(protein = if_else(is.na(protein), 0, protein),
           transcript = if_else(is.na(transcript), 0, transcript)) %>%
    filter(!(pinetree %in% IGNORE_GENES), !is.na(gene_number))
}

data <- load_sim("../data/simulation/phage/phage_wildtype.tsv")

bar_sim_plot <- ggplot(data %>% filter(time == 1000), aes(x=factor(gene_number, gene_labels), 
                                                             y=transcript, fill=factor(class))) +
  geom_bar(stat = "identity") +
  geom_vline(xintercept = (match(c("3.8", "4.7", "6.5", "18.5"), gene_labels)) - 0.5, 
             linetype="dashed", color = "gray40") +
  geom_vline(xintercept = (match(c("11"), gene_labels)) - 0.5) +
  theme(legend.position = "none") +
  labs(x="gene", y="simulated transcripts") +
  theme(axis.text.x=element_blank(), axis.ticks.x=element_blank()) +
  scale_y_continuous(expand = c(0,0)) +
  scale_fill_manual(values=cbPalette)

data_degrade <- load_sim("../data/simulation/phage/phage_rnase_cleavage_counts.tsv")

bar_sim_degrade <- ggplot(data_degrade %>% filter(time == 1000), aes(x=factor(gene_number, gene_labels), 
                                                             y=transcript, fill=factor(class))) +
  geom_bar(stat = "identity") +
  geom_vline(xintercept = (match(c("3.8", "4.7", "6.5", "18.5"), gene_labels)) - 0.5, 
             linetype="dashed", color = "gray40") +
  geom_vline(xintercept = (match(c("11"), gene_labels)) - 0.5) +
  theme(legend.position = "none") +
  labs(x="gene", y="simulated transcripts") +
  theme(axis.text.x=element_blank(), axis.ticks.x=element_blank()) +
  scale_y_continuous(expand = c(0,0)) +
  scale_fill_manual(values=cbPalette)

rna_9_dist_rnase <- rna_9_dist + 
  geom_vline(xintercept = (match(c("3.8", "4.7", "6.5", "18.5"), gene_labels)) - 0.5, 
             linetype="dashed", color = "gray40")

sim_exp_degrad_9_plot <- plot_grid(rna_9_dist_rnase, bar_sim_plot, bar_sim_degrade,
                                   align = 'v', ncol = 1, labels = "AUTO")

lineplot <- ggplot(data_degrade, aes(x=time, y=transcript, group = gene_number, color=factor(class))) +
  geom_line() +
  scale_y_continuous(trans = "log10") + 
  xlab("time (s)") +
  theme(legend.title = element_blank(), legend.spacing.x = unit(2, 'pt')) + 
  scale_color_manual(values=cbPalette, breaks=c("1","2","3"), labels=c("class I", "class II", "class III")) + xlim(0,1000) 


if (SAVE_FIGS) {
  save_plot("./figures/lineplot.pdf", lineplot, base_height = 4, base_aspect_ratio = 1.7)
  save_plot("./figures/T7_rna_exp_sim_degrade_no_labels.pdf", sim_exp_degrad_9_plot, 
            base_width = 8, base_height = 9)
}

```

### Simulated promoter knockouts and codon deoptimization

```{r}
wildtype_sim <- load_sim("../data/simulation/phage/phage_wildtype.tsv") %>% mutate(strain = "wildtype")
recoded_sim <- load_sim("../data/simulation/phage/phage_model_recoded_27.tsv") %>% mutate(strain = "recoded")
ko_sim <- load_sim("../data/simulation/phage/phage_model_knockout_9_10_27.tsv") %>% mutate(strain = "knockout")
ko_recoded_sim <- load_sim("../data/simulation/phage/phage_model_knockout_9_10_recoded.tsv") %>% mutate(strain = "knockout_recoded")

sims <- bind_rows(wildtype_sim, recoded_sim, ko_sim, ko_recoded_sim)

sims <- sims %>%
  filter(time == 1000, gene_number %in% c("7", "8", "9", "10A", "11", "12")) %>%
  select(gene_number, protein, strain)

bar_plot <- ggplot(sims, aes(x = factor(gene_number, gene_labels), y = protein, fill = gene_number)) +
  geom_bar(stat = "identity") +
  facet_wrap( . ~ strain) +
  xlab("gene") + 
  ylab("protein abundance")

sims_relative <- sims %>% 
  filter(gene_number == "10A")

strain_labels = c("wildtype" = "T7.61", 
                  "recoded"=expression(paste("gene ", italic("10"), " recoded")), 
                  "knockout"= expression(paste(Delta*phi, "9/10")), 
                  "knockout_recoded"=expression(paste(Delta*phi, "9/10, gene ", italic("10"), " recoded")))
strain_levels = c("knockout_recoded",
                  "knockout",
                  "recoded", 
                  "wildtype")

# These fitness values come from taking gene 10A protein abundance/~400 proteins per capsid,
# then solving for intrinsic growth rate (r) as described by Bull, et al 2011 using wolfram alpha. 
# Then we transform r into doublings per hour dblgs = log2(e^(r*t)), where t = 60 min (1hr).
## wildtype - 24
## recoded - 20.88
## knockout - 13.7
## knockout/recoded - 11.07
## 
## For the simulations with degradation:
## wildtype - 27.7
## recoded - 26.21
## knockout - 21.98
## knockout/recoded -

sims_transformed <- data_frame(strain=c('wildtype', 'recoded', 'knockout', 'knockout_recoded'),
                               fitness=c(27.7, 26.21, 21.98, 20.68),
                               type=c('sim', 'sim', 'sim', 'sim'))

fit <- read_xlsx("../data/fitness.xlsx")

exp_fit <- fit %>% filter(!grepl('evo', Strain), 
                            !grepl('L2', Strain), 
                            Background != "8st",
                            strain_rna %in% c("T7Hi", "phi910v2", "11-44", "11-44-phi910v2")) %>%
  select(strain = strain_rna, fitness = Fitness) %>%
  mutate(type = 'exp',
         strain = case_when(
           strain == "11-44-phi910v2" ~ "knockout_recoded", 
           strain == "phi910v2" ~ "knockout", 
           strain == "11-44" ~ "recoded", 
           strain == "T7Hi" ~ "wildtype"
         )) %>%
  group_by(strain, type) %>% 
  summarize(fitness = mean(fitness))

exp_sim_fit <- bind_rows(exp_fit, sims_transformed) %>%
  spread(strain, fitness) %>%
  mutate_at(vars(knockout:wildtype), funs(. / wildtype)) %>%
  gather(strain, fitness, -type)

exp_sim_fit_plot <- ggplot(exp_sim_fit, aes(x = factor(strain, levels=strain_levels), y = fitness, fill=factor(type, levels=c('sim', 'exp')))) + 
  geom_bar(stat = "identity", alpha=0.7, position = "dodge") +
  scale_y_continuous(expand = c(0,0)) +
  xlab("strain") + ylab(expression(paste("fitness relative to T7.61"))) +
  scale_x_discrete(labels=strain_labels) + 
  scale_fill_manual(name="type", values=cbPalette[c(6,7)], labels=c("exp"="experimental", "sim"="simulated"), breaks=c('exp', 'sim')) + coord_flip()



if(SAVE_FIGS == TRUE) {
  save_plot("./figures/recoded_knockout.pdf", exp_sim_fit_plot, base_aspect_ratio = 2)
}

```


### Proteomics and rna correlations

```{r}

proteomics <- read_csv("../output/proteomics.csv")

proteomics <- proteomics %>%
  mutate(strain = str_replace_all(strain, c("11-44" = "atten", "11-42" = "evol", "11-46" = "wt"))) %>%
  mutate(protein_desc = trimws(protein_desc))
  
# Join in standard T7 gene labels.
proteomics <- filter(proteomics, org == "phage") %>% 
  group_by(strain, time, b_rep) %>%
  # Join by each group seperately so that missing phage proteins are filled in with NAs
  do(full_join(., labels, by = c("protein_desc" = "accession")) %>% fill(strain, time, b_rep)) %>%
  mutate(area_norm = ifelse(is.na(area_norm), 0, area_norm))
         # gene_number = str_replace_all(gene_number, c("10A" = "10"))) # Fill in missing proteins (NAs) with 0

proteomics <- proteomics %>%
  group_by(time, strain, gene_number, class, protein_desc) %>% 
  filter(strain == "wt", area_norm != 0) %>%
  group_by(gene_number, class, time) %>%
  summarize(area_norm = mean(area_norm)) %>%
  ungroup() %>% 
  group_by(time) %>% 
  mutate(area_total = sum(area_norm)) %>% 
  group_by(gene_number, class, time) %>% 
  mutate(area_norm = area_norm/area_total)

phage_rna <- transcripts %>% 
  group_by(time, gene_number) %>% 
  summarize(avg_tpm = mean(tpm)/1000000) %>% 
  filter(avg_tpm != 0)

proteomics_rna <- inner_join(proteomics, phage_rna, by = c("gene_number" = "gene_number", "time" = "time"))

time_labels <- c('5' = '5 minutes', '9' = '9 minutes')

prot_rna_relabel <- proteomics_rna %>%
  mutate(new_class = case_when(
    class == 2 && gene_number %in% c(1.2, 1.3, 1.6, 1.7, 1.8, 2, 2.5, 2.8, 3, 3.5, 3.8) ~ "2A",
    class == 3 ~ "3",
    class == 1 ~ "1",
    TRUE ~ "2B")
  )

exp_cor_relabel <- ggplot(prot_rna_relabel %>% filter(time != 1), 
                  aes(x = log10(avg_tpm), y = log10(area_norm), color = factor(new_class), shape = factor(new_class))) + 
  geom_point() + 
  xlab("log10(relative transcript abundance)") + 
  ylab("log10(relative protein abundance)") + 
  scale_color_manual(values=cbPalette,  breaks=c("1","2","3"), labels=c("class I", "class II", "class III")) + 
  theme(legend.title = element_blank()) + 
  scale_shape_discrete() +
  facet_wrap(~time, scales = "free", labeller = as_labeller(time_labels)) +
  stat_ellipse() + 
  scale_x_continuous(limits = c(-3.4, -0.8)) + 
  scale_y_continuous(limits = c(-3.7, -0.4)) + theme(legend.position = c(0.92,0.2))

exp_cor <- ggplot(proteomics_rna %>% filter(time != 1), 
                  aes(x = log10(avg_tpm), y = log10(area_norm), color = factor(class), shape = factor(class))) + 
  geom_point() + 
  xlab("log10(relative transcript abundance)") + 
  ylab("log10(relative protein abundance)") + 
  scale_color_manual(values=cbPalette, breaks=c("1","2","3"), labels=c("class I", "class II", "class III")) + 
  scale_shape_discrete(breaks=c("1","2","3"), labels=c("class I", "class II", "class III")) +
  facet_wrap(~time, scales = "free", labeller = as_labeller(time_labels)) +
  stat_ellipse() + 
  scale_x_continuous(limits = c(-3.4, -0.8)) + 
  scale_y_continuous(limits = c(-3.7, -0.4)) + theme(legend.position = c(0.0,0.15), legend.title = element_blank())

if(SAVE_FIGS == TRUE) {
  save_plot("./figures/prot_rna_scatter.pdf", exp_cor, base_width = 7, base_height = 3.5)
}

# Compute correlations between RNA and protein

proteomics_rna %>%
  filter(time != 1, class != 1) %>%
  group_by(time, class) %>%
  summarize(cor = cor.test(log10(area_norm), log10(avg_tpm))$estimate, p.value = cor.test(log10(area_norm), log10(avg_tpm))$p.value)

```

## Simulated proteomics and RNA correlations

```{r}

bar_sim_relabel <- bar_sim %>%
  mutate(new_class = case_when(
    class == 2 && gene_number %in% c('1.2', '1.3', '1.4', '1.5', '1.6', '1.7', '1.8', '2', '2.5', '2.8', '3', '3.5', '3.8') ~ "2A",
    class == 3 ~ "3",
    class == 1 ~ "1",
    TRUE ~ "2B")
  )

data_degrade_renorm <- data_degrade %>%
  group_by(time) %>%
  mutate(total_prot = sum(protein), total_transcript = sum(transcript)) %>%
  mutate(protein = protein/total_prot, transcript = transcript/total_transcript)

simulated_cor <- ggplot(data_degrade_renorm %>% filter(time == 500 | time == 1000), aes(x = log10(transcript), y = log10(protein), color = factor(class), shape = factor(class))) + 
  geom_point() +
  xlab("log10(simulated transcript abundance)") + 
  ylab("log10(simulated protein abundance)") + 
  scale_color_manual(values=cbPalette, breaks=c("1","2","3"), labels=c("class I", "class II", "class III")) +
  scale_shape_discrete(breaks=c("1","2","3"), labels=c("class I", "class II", "class III")) +
  facet_wrap(~time, scales="free") +
  stat_ellipse() + 
 theme(legend.position = c(0.86,0.15), legend.title=element_blank())

scatter_plots <- plot_grid(exp_cor, simulated_cor, nrow = 2, labels="AUTO", align='V')

if(SAVE_FIGS == TRUE) {
  save_plot("./figures/sim_scatter_plots.pdf", scatter_plots, base_width = 8, base_height = 8)
}

data_degrade_renorm %>%
  filter(time == 500 | time == 1000, class != 1, protein > 0, transcript > 0) %>%
  group_by(time, class) %>%
  summarize(cor = cor.test(log10(transcript), log10(protein))$estimate, p.value = cor.test(log10(transcript), log10(protein))$p.value)

```

## Experimental and simulated protein distributions

```{r}
exp_prot_final <- proteomics %>% 
  filter(!is.na(gene_number)) %>% 
  ungroup() %>% 
  mutate(gene_number = str_replace(gene_number, "10", "10A"))

prot_genes <- unique(exp_prot_final$gene_number)

sim_plot_prot <-ggplot(sim_final %>% filter(gene_number %in% prot_genes), 
                       aes(x=factor(gene_number, unique(mixedsort(gene_number))), y=proteins, fill= factor(class))) + 
  geom_bar(stat="identity") + xlab("gene") + ylab("simulated protein abundance") +
  theme(axis.text.x=element_blank(), axis.ticks.x=element_blank(), legend.position = c(0.0, 0.85)) +
  scale_fill_discrete(name="class") + scale_y_continuous(expand = c(0, 0))

exp_plot_prot <- ggplot(exp_prot_final, 
                        aes(x = factor(gene_number, levels=unique(mixedsort(gene_number))), y = area_norm, fill = factor(class))) + 
  geom_bar(stat="summary", fun.y = "mean") +
  geom_point(alpha=0.5) +
  theme(axis.text.x=element_blank(), axis.title.x=element_blank(), axis.ticks.x=element_blank(), legend.position = c(0.0, 0.85)) +
  scale_color_discrete("replicate") +
  ylab("relative protein abundance") +
  scale_fill_discrete(name="class") + scale_y_continuous(expand = c(0, 0))

sim_exp_plot_prot <- plot_grid(exp_plot_prot, sim_plot_prot, cols = 1, align = "v")

save_plot("T7_prot_exp_sim_no_labels.pdf", sim_exp_plot_prot, base_width = 8, base_height = 6)
```