---
title: "TRNA Protein Analysis"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE)

library(tidyverse)
library(ggplot2)
library(gtools)
library(stringr)
library(cowplot)
library(readxl)
SAVE_FIGS <- FALSE

cbPalette <- c("#CC79A7", "#56B4E9", "#E69F00", "#F0E442", "#0072B2", "#D55E00", "#009E73", "#999999")

# These are genes that are not included in the simulation and should be removed before comparing
# simulations with experimental data
IGNORE_GENES <- c("gene 10B",
                "possible gene 5.5-5.7",
                "gene 1.5",
                "gene 1.6",
                "gene 4.1",
                "gene 4B",
                "gene 0.6A",
                "gene 0.6B",
                "possible gene 0.6B",
                "gene 0.5",
                "gene 0.4")
```

```{r}
id_map <- read_csv("id_map.csv") %>% filter(!is.na(gene_number), !(pinetree %in% IGNORE_GENES))
gene_labels <- unique(mixedsort(id_map$gene_number))
```

# Recreate the plot of transcripts per million 9 min postinfection (fig. 1B) from Jack et al. 2019

```{r}
transcription_data <- "transcripts_37C.csv" # measured transcript abundances downloaded from https://github.com/benjaminjack/phage_simulation

transcripts <- read_csv(transcription_data) %>% 
  filter(!is.na(class), !(pinetree %in% IGNORE_GENES))

gene_labels <- unique(mixedsort(id_map$gene_number))

rna_9_dist <- ggplot(transcripts %>% filter(time == 9), aes(x = factor(gene_number, levels=gene_labels), y = tpm)) + 
  geom_bar(stat="summary", fun.y = "mean", aes(fill = factor(class))) +
  geom_point(alpha = 0.5) + 
  scale_fill_manual(values=cbPalette, breaks=c("1","2","3"), labels=c("class I", "class II", "class III")) +
  xlab("T7 gene") + ylab("Transcripts per million") + scale_y_continuous(expand = c(0,0)) +
  coord_cartesian(clip = "off") +
  theme_half_open()
```

Read in data for WT static model

```{r}
fop <- 0.7 # near WT fop
# Final codon speeds are 7 and 3 for opt and non-opt codons, respectively

labels <- read_csv("id_map.csv") %>% filter(!is.na(gene_number))
load_sim <- function(path) {
  read_tsv(path) %>% group_by(time) %>% nest() %>%
    mutate(data = purrr::map(data, ~left_join(labels, ., by = c("pinetree" = "species")))) %>%
    unnest() %>%
    mutate(time = round(time)) %>%
    mutate(protein = if_else(is.na(protein), 0, protein),
           transcript = if_else(is.na(transcript), 0, transcript)) %>%
    filter(!(pinetree %in% IGNORE_GENES), !is.na(gene_number))
}

base_dir <- "../data/simulation/revised_weighted_opt7.0_nonopt3.0/"
seeds <- c(70, 38, 79, 4, 80, 84, 27, 33, 7, 51)
weighted_data_07 <- NULL
for (seed in seeds) {
  tmp <- load_sim(paste0(base_dir,"weighted_opt7.0_nonopt3.0_",seed,"_fop",fop,".tsv"))
  tmp <- tmp %>% filter(time == 1200) %>%mutate(seed = seed)
  if (is.null(weighted_data_07)) {
    weighted_data_07 <- tmp
  }
else {
    weighted_data_07 <- rbind(weighted_data_07, tmp)
  }
}
```

```{r}
weighted_data_07_ave <- weighted_data_07 %>% 
  group_by(gene_number, class) %>% 
  summarize(mean_protein = mean(protein),
            std_protein = sd(protein)) %>%
  ungroup()
```


```{r}
static_model <- ggplot(weighted_data_07_ave%>%mutate(std_err = std_protein/sqrt(10)), 
                       aes(x=factor(gene_number, gene_labels), y=mean_protein, fill=factor(class))) +
  geom_bar(stat = "identity") +
  geom_errorbar(aes(ymin = mean_protein-std_err, 
                    ymax = mean_protein+std_err), 
                width=0.02,
                alpha=0.5,
                linewidth=1) +
  labs(x="T7 gene", y="Protein abundance") +
  theme(axis.text.x=element_blank(), axis.ticks.x=element_blank()) +
  scale_y_continuous(expand = c(0,0)) +
  scale_fill_manual(values=cbPalette) +
  theme_half_open()

static_model
```

Read in data for WT dynamic model

```{r}
fop <- 0.7
# final parameters are charging rate of 0.5, ribosome speed of 1.5, 50000 total tRNAs, and 7:3 tRNA ratio

base_dir <- "../data/simulation/revised_dynamic_pref0.7_charge0.5_ribospeed1.5/"
seeds <- c(70, 38, 79, 4, 80, 84, 27, 33, 7, 51)
dynamic_data_07 <- NULL
for (seed in seeds) {
  tmp <- load_sim(paste0(base_dir,"trna_phage_pref0.7_",seed,"_fop",fop,"_ribospeed1.5_trna50000.tsv"))
  tmp <- tmp %>% filter(time == 1200) %>%mutate(seed = seed)
  if (is.null(dynamic_data_07)) {
    dynamic_data_07 <- tmp
  }
  else {
    dynamic_data_07 <- rbind(dynamic_data_07, tmp)
  }
}

dynamic_data_07_ave <- dynamic_data_07 %>% 
  group_by(gene_number, class) %>% 
  summarize(mean_protein = mean(protein),
            std_protein = sd(protein))

```

```{r}
dynamic_model <- ggplot(dynamic_data_07_ave%>%mutate(std_err = std_protein / sqrt(10)), aes(x=factor(gene_number, gene_labels), 
                                                             y=mean_protein, fill=factor(class))) +
  geom_bar(stat = "identity") +
  geom_errorbar(aes(ymin = mean_protein-std_err, 
                    ymax = mean_protein+std_err), 
                width=0.02,
                alpha=0.5,
                linewidth=1) +
  labs(x="T7 gene", y="Protein abundance") +
  theme(axis.text.x=element_blank(), axis.ticks.x=element_blank()) +
  scale_y_continuous(expand = c(0,0)) +
  scale_fill_manual(values=cbPalette) +
  theme_half_open()
```

Combine all three distribution plots

```{r}
gene_dis_theme <-  theme(axis.text.x=element_blank(), 
        axis.ticks.x=element_blank(), 
        #legend.position = c(0.0, 0.8), 
        legend.title = element_blank(), 
        legend.spacing.x = unit(2, 'pt'),
        panel.grid = element_blank(),
        plot.title = element_text(size=12))

legend <- get_legend(rna_9_dist+gene_dis_theme)

c1 <- cowplot::plot_grid(rna_9_dist+gene_dis_theme+theme(legend.position = "none"), 
                   static_model+gene_dis_theme+theme(legend.position = "none"), 
                   dynamic_model+gene_dis_theme+theme(legend.position = "none"), 
                   ncol = 1,
                   align = "vh", labels = c("A", "B", "C"))

cowplot::plot_grid(c1, legend, rel_widths = c(1, 0.1))
ggsave("fig/gene_dist_both_models.pdf", height = 8, width = 10)
```


