---
title: "Fraction of Optimal Codons x Fitness - Revised Weighted & Dynamic"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE)

library(tidyverse)
library(ggplot2)
library(gtools)
library(stringr)
library(cowplot)
library(readxl)
SAVE_FIGS <- TRUE

cbPalette <- c("#CC79A7", "#56B4E9", "#E69F00", "#F0E442", "#0072B2", "#D55E00", "#009E73", "#999999")

# These are genes that are not included in the simulation and should be removed before comparing
# simulations with experimental data
IGNORE_GENES <- c("gene 10B",
                "possible gene 5.5-5.7",
                "gene 1.5",
                "gene 1.6",
                "gene 4.1",
                "gene 4B",
                "gene 0.6A",
                "gene 0.6B",
                "possible gene 0.6B",
                "gene 0.5",
                "gene 0.4")
```

```{r}
max_time <- 1200
gene_name <- "10A"
fops<- c(0.0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7)
#fops <- c(0.1, 0.7)
seeds <- c(70, 38, 79, 4, 80, 84, 27, 33, 7, 51)

########## Weighted Model
opt <- 7
nonopt <- 3
weighted_output_dir <- sprintf("../data/simulation/revised_weighted_opt%.1f_nonopt%.1f", opt, nonopt)

load_sim <- function(path) {
  read_tsv(path) %>% group_by(time) %>%
    mutate(time = round(time))
}

w_data <- NULL
for (seed in seeds) {
  seed_data <- NULL
    for (fop in fops) {
      filename <- sprintf("weighted_opt%.1f_nonopt%.1f_%d_fop%.1f.tsv", opt, nonopt, seed, fop)
      tmp <- load_sim(paste0(weighted_output_dir,"/",filename))
      tmp <- tmp %>%mutate(fop = fop)
      if (is.null(seed_data)) {
        seed_data <- tmp
      }
      else {
        seed_data <- rbind(seed_data, tmp)
      }
    }  
    seed_data <- seed_data %>% mutate(seed = seed)
    if (is.null(w_data)) {
        w_data <- seed_data
    }
    else {
      w_data <- rbind(w_data, seed_data)
    }
}
```

```{r}
# Convert 10A protein abundances to fitness values (doubling per hour)
w_data <- w_data %>%
  filter(species == "gene 10A", time == 1200) %>%
  mutate(
    r = map_dbl(protein, ~ {
      f <- function(r) ((.x / 400) * exp(-12 * r) - 1)
      res <- tryCatch(uniroot(f, lower = -1, upper = 1)$root, error = function(e) NA)
      return(res)
    }),
    fitness = log2(exp(60 * r)),  # Compute doublings per hour
  )

# Normalize fitness values
w_data <- w_data %>%
  group_by(fop) %>%  
  summarize(mean_fitness = mean(fitness), 
            std_fitness = sd(fitness),
            mean_protein = mean(protein),
            std_protein = sd(protein)) %>%
  ungroup() %>%
  mutate(mean_rescale = mean_fitness / max(mean_fitness)) %>%
  mutate(std_rescale = std_fitness / max(mean_fitness))
```

```{r}
setwd("/stor/scratch/Wilke/amh7958/tRNA_dynamics")

########## Dynamic Model
pref_val <- 0.7
charge_val <- 0.5
ribo_speed <- 1.5
trna_count <- 50000
fops <- seq(0, 0.7, by = 0.1)
seeds <- c(70, 38, 79, 4, 80, 84, 27, 33, 7, 51)
dynamic_output_dir <- sprintf("data/simulation/revised_dynamic_pref%.1f_charge%.1f_ribospeed%.1f", pref_val, charge_val, ribo_speed)

# Generate file paths
file_paths <- expand.grid(seed = seeds, fop = fops, pref_val = pref_val, ribo_speed = ribo_speed, trna_count = trna_count) %>%
  mutate(output_dir = dynamic_output_dir) %>%
  mutate(file_name = sprintf("trna_phage_pref%.1f_%d_fop%.1f_ribospeed%.1f_trna%d.tsv", pref_val, seed, fop, ribo_speed, trna_count)) %>%
  mutate(full_path = file.path(output_dir, file_name))

labels <- read_csv("R/id_map.csv") %>% filter(!is.na(gene_number))
load_sim <- function(paths) {
  data_list <- lapply(paths, function(path) {   # loop through each file path in paths
    read_tsv(path) %>%   # read the tsv file at the given path
      mutate(file = path) %>%   # add column 'file' for file path
      group_by(time) %>%   # group by time
      nest() %>%   # convert each group in a nested data frame of 2 columns: time and data (all the rows for that time point)
      mutate(data = purrr::map(data, ~full_join(labels, ., by = c("pinetree" = "species")))) %>%   # join with the labels data frame
      unnest(cols = c(data)) %>%   # expand the nested data back into a flat data frame
      mutate(time = round(time)) %>%   # round the time column
      mutate(protein = if_else(is.na(protein), 0, protein),   # replace missing protein and transcript values with 0
             transcript = if_else(is.na(transcript), 0, as.numeric(transcript))) %>%
      filter(!(pinetree %in% IGNORE_GENES), !is.na(gene_number))   # filter out unwanted genes
  })
  bind_rows(data_list)   # combine data frames in data_list into a single data frame
}

# Load all simulations
dynamic_data <- load_sim(file_paths$full_path)

# Add seed and fop columns and filter to include only the specified gene and max time
dynamic_data <- dynamic_data %>%
  mutate(
    seed = as.numeric(str_extract(file, "(?<=_)[0-9]+(?=_fop)")),
    fop = as.numeric(str_extract(file, "(?<=fop)[0-9]+\\.[0-9]")),
  ) %>%
  filter(time == max_time, str_detect(pinetree, paste("gene", gene_name)))

# Convert 10A protein abundances to fitness values (doubling per hour)
dynamic_data <- dynamic_data %>%
  mutate(
    r = map_dbl(protein, ~ {
      f <- function(r) ((.x / 400) * exp(-12 * r) - 1)
      res <- tryCatch(uniroot(f, lower = -1, upper = 1)$root, error = function(e) NA)
      return(res)
    }),
    fitness = log2(exp(60 * r)),  # Compute doublings per hour
  )

# Normalize fitness values
dynamic_data <- dynamic_data %>%
  group_by(fop) %>%  
  summarize(mean_fitness = mean(fitness), 
            std_fitness = sd(fitness),
            mean_protein = mean(protein),
            std_protein = sd(protein)) %>%
  ungroup() %>%
  mutate(mean_rescale = mean_fitness / max(mean_fitness)) %>%
  mutate(std_rescale = std_fitness / max(mean_fitness))
```

```{r}
########## Combine Data

# Combine weighted and dynamic data into one data frame
combined_data <- bind_rows(
  dynamic_data %>% mutate(model = "Dynamic"),
  w_data %>% mutate(model = "Weighted")
)

# Filter data to include only fop values from 0.7 to 0
combined_data <- combined_data %>%
  filter(fop <= 0.7) %>%
  mutate(type = "Simulation")
```


```{r}
n <- 10 # number of simulation replicates
palette <- c("#B63441", "#40476D")

plot_model_fitness <- ggplot(combined_data%>%
                               mutate(std_err_protein = std_protein / sqrt(n)) %>%
                               filter(fop >= 0), 
                        aes(x = fop, y = mean_protein, color=model)) +
  geom_line(linewidth = 1) +
  geom_errorbar(aes(ymin = mean_protein-std_err_protein, 
                    ymax=mean_protein+std_err_protein), 
                width=0.02,
                alpha=0.5,
                linewidth=1) +
  # geom_smooth(method = "gam", formula = y ~ s(x, bs = "cs"), se = FALSE, size = 0.75) +
  #geom_smooth(method = "lm", se = FALSE, size = 0.75) +
  geom_point(size = 3, aes(fill=model), color = "white", shape=21)+
  labs(x = expression(f[op]~"(gene 10A)"), y = "Protein 10A abundance") +
  theme_bw() +
  scale_color_manual(values = c("Weighted" = palette[2], "Dynamic" = palette[1]),
                     labels = c("Weighted" = "Static tRNA model", "Dynamic" = "Dynamic tRNA model"),
                     name="",
                     guide = guide_legend(reverse = TRUE)) +
  scale_fill_manual(values = c("Weighted" = palette[2], "Dynamic" = palette[1]), guide="none") +
  scale_x_continuous(breaks = seq(0, 1.0, by = 0.1), labels = seq(0, 1.0, by = 0.1), trans = "reverse") +
  theme(text = element_text(size=12))

plot_model_fitness
#ggsave("alexis_fig/10A_protein_simulation_only.pdf", width=7, height=4)
```

```{r}
# create dataset with T7 fitness data from Bull 2012 (fig 1)
bull_data <- data.frame(fop = c(0.68, 0.5, 0.3, 0.2, 0.1),
                              mean_fitness = c(43.21, 40.365, 39.1, 36.1, 35.665),
                              std_fitness = c(0.3346, 0.415, 0.435, 0.41, 0.762))
bull_data <- bull_data %>% 
  mutate(mean_rescale = mean_fitness / max(mean_fitness)) %>%
  mutate(std_rescale = std_fitness / max(mean_fitness) * 10) %>%
  mutate(type = "Experiment")

# make duplicates for each facet panel
bull_data_dynamic <- bull_data %>% mutate(model = "Dynamic")
bull_data_weighted <- bull_data %>% mutate(model = "Weighted")

combined_data_bull <- bind_rows(combined_data, bull_data_dynamic, bull_data_weighted)
```

```{r}
combined_data_bull$model <- factor(combined_data_bull$model, levels = c("Weighted", "Dynamic"))
model_color <- "#B63441"
linewidth <- 1
facet_labels <- c("Weighted" = "Static tRNA model", "Dynamic" = "Dynamic tRNA model")

ggplot(combined_data_bull%>% mutate(std_err = std_rescale / sqrt(n)) %>% filter(fop > 0), 
       aes(x = fop, y = mean_rescale, color=type)) +
  facet_wrap(~model, labeller = as_labeller(facet_labels), ncol = 1) +
  geom_line(linewidth=linewidth) +
  geom_errorbar(aes(ymin = mean_rescale-std_err, 
                    ymax=mean_rescale+std_err), 
                width=0.02,
                alpha=0.5,
                linewidth=linewidth) +
  geom_point(size = 3, aes(fill=type), color="white", shape=21) +
  # adjust theme and axis labels
  scale_x_continuous(breaks = seq(0, 1.0, by = 0.2), 
                     labels = seq(0, 1.0, by = 0.2), 
                     trans="reverse",
                     name=expression(f[op]~"(gene 10A)")) +
  scale_y_continuous(name="Normalized T7 fitness") +
  scale_color_manual(values = c("grey55", model_color),
                     name="",
                     guid=guide_legend(reverse = TRUE)) +
  scale_fill_manual(values = c("grey45", model_color), guide="none") +
  theme_bw() +
  theme(text = element_text(size = 12)) 
  
#ggsave("alexis_fig/fitness_v2.pdf", width=6, height=5)
```

```{r}
n <- 10
linewidth <- 1
facet_labels <- c("Weighted" = "Static tRNA model", "Dynamic" = "Dynamic tRNA model")
palette <- c("#40476D", "#B63441")

fitness_plot_model <- ggplot(combined_data_bull%>%filter(type=="Simulation")%>%mutate(std_err = std_rescale / sqrt(n)) %>% filter(fop >= 0), 
       aes(x = fop, y = mean_rescale, color=model)) +
  geom_line(linewidth=linewidth) +
  geom_errorbar(aes(ymin = mean_rescale-std_err, 
                    ymax=mean_rescale+std_err), 
                width=0.02,
                alpha=0.5,
                linewidth=linewidth) +
  geom_point(size = 3, aes(fill=model), color="white", shape=21) +
  # adjust theme and axis labels
  scale_x_continuous(breaks = seq(0, 1.0, by = 0.2), 
                     labels = seq(0, 1.0, by = 0.2), 
                     trans="reverse",
                     name=expression(f[op]~"(gene 10A)")) +
  scale_y_continuous(name="Predicted fitness", 
                     limits=c(0.7, 1.05)) +
  scale_color_manual(values = c("Weighted" = palette[1], "Dynamic" = palette[2]),
                     labels = c("Weighted" = "Static tRNA model", "Dynamic" = "Dynamic tRNA model"),
                     name="",
                     guid=guide_legend(reverse = FALSE)) +
  scale_fill_manual(values = palette, guide="none") +
  theme_bw() +
  theme(text = element_text(size = 12),
        legend.position = "inside",
        legend.position.inside = c(0.3,0.2),
        legend.title = element_blank()
  )
```

```{r}
fitness_plot_experiment <- ggplot(combined_data_bull %>%
                                    filter(type=="Experiment") %>%
                                    filter(model=="Dynamic") %>%
                                    mutate(std_rescale = std_rescale / 10) %>%
                                    mutate(std_err = std_rescale) %>% 
                                    filter(fop >= 0), 
       aes(x = fop, y = mean_rescale)) +
  #geom_line(linewidth=linewidth) +
  geom_smooth(linewidth=linewidth, method="lm", se=FALSE, color="grey55", fullrange=TRUE) +
  geom_errorbar(aes(ymin = mean_rescale-std_err, 
                    ymax=mean_rescale+std_err), 
                width=0.02,
                alpha=0.5,
                linewidth=linewidth) +
  #geom_point(size = 3, fill="grey55", color="white", shape=21) +
  geom_point(size = 3, color="grey55") +
  # adjust theme and axis labels
  scale_x_continuous(trans="reverse",
                     limits=c(0.7, 0),
                     name=expression(f[op]~"(gene 10A)")) +
  scale_y_continuous(name="Measured fitness",
                     limits=c(0.7, 1.05)) +
  theme_bw() +
  theme(text = element_text(size = 12),
        )
```

```{r}
cowplot::plot_grid(fitness_plot_model, fitness_plot_experiment, labels=c("A", "B"))
ggsave("fig/fitness_revised.pdf", width = 8, height = 3.5)
```

```{r}
# tRNA abundances as a function of time for fop=0.7

dir <- "../data/simulation/revised_dynamic_pref0.7_charge0.5_ribospeed1.5_timestep_5/"
fops <- c(0.7, 0.1)

dynamic_timecourse <- NULL
for (fop in fops) {
  tmp <- read_tsv(paste0(dir, "trna_phage_pref0.7_1", "_fop", fop, "_ribospeed1.5_trna50000.tsv"))
  tmp <- tmp %>% 
    mutate(time = round(time)) %>% 
    mutate(fop = fop)
  if (is.null(dynamic_timecourse)) {
    dynamic_timecourse <- tmp
  }
  else {
    dynamic_timecourse <- rbind(dynamic_timecourse, tmp)
  }
}
```

```{r}
color <- "#0090C1"
facet_labels <- c("0.1" = "Deoptimized", "0.7" = "Wild-type")

dynamic_timecourse %>%
  filter(species %in% c("pref_charged", "nonpref_charged")) %>%
  filter(fop %in% c(0.1, 0.7)) %>%
  mutate(fop = factor(fop, level=c(0.7, 0.1))) %>%
  ggplot(aes(x=time, y=protein, color=species)) +
  facet_wrap(~fop, labeller = as_labeller(facet_labels)) +
  geom_line(linewidth=1.2) +
  scale_x_continuous(name="Time (s)",
                     limits = c(0, 1200)) +
  scale_y_continuous(name="Count",
                     limits = c(0, 40000)) +
  scale_color_manual(name="",
                     values = c("pref_charged" = "black", "nonpref_charged" = color),
                     breaks=c("pref_charged", "nonpref_charged"),
                     labels = c(expression(T[c1]), expression(T[c2]))) +
  geom_vline(xintercept = 440, alpha = 0.3) +
  geom_vline(xintercept = 364, alpha = 0.3) +
  geom_vline(xintercept = 250, alpha = 0.3) +
  theme_bw() +
  theme(text = element_text(size = 12))

#ggsave("alexis_fig/t7_dynamic_model_trna_vs_time_wt_deop.pdf", width = 8, height = 3.5)
```


